### PR TITLE
Java+DB docs [skip ci]

### PR DESCRIPTION
**Pull request by boczkowska**
**What type of PR is this?**
/kind documentation
[skip ci]

**What does this PR do / why we need it**:
Adds Java+DB on minikube guide to docs, based on @dharmit doc https://github.com/dharmit/application-stack-samples/blob/minikube-readme/jpa/README-minikube.md 
**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/4116

**PR acceptance criteria**:

- [ ] Unit test 

- [ ] Integration test 

- [x] Documentation 

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
I edited the original doc slightly to match the general structure of the docs and made a few changes to the style/wording.

### CODE REVIEW

File: docs/public/deploying-java-app-with-database.adoc [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+** <your-email> is your Docker email.
+
+* Add this new cred ('regcred' in the example above) to the default service account in minikube:
++
+[source,sh]
+----
+$ kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "regcred"}]}'
+----
+
+Kubernetes Dashboard graphical UI config:<br>
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:110]: This `<br>` shows as-is even in the rich diff view of PR. I hope it won't show up in the docs.

**Comment by dharmit** [comment]
My concern with the detailed steps about setting up minikube, using docker driver, using pull secret for Docker Hub rate limit is that do we want to cover this on odo.dev website?

My assumption while writing the docs so far has been that user has a CRC/minikube instance setup and working. Although I do agree that enabling ingress and olm add-ons is essential here.

TBH, I'm confused about the amount of detail we want to go into in the docs. 

@kadel what are your thoughts?

**Comment by kadel** [comment]
> My concern with the detailed steps about setting up minikube, using docker driver, using pull secret for Docker Hub rate limit is that do we want to cover this on odo.dev website?

It depends on how common the given problem is. If it is something that almost every user will face, and it is a non-standard path we should document it.

**Comment by dharmit** [comment]
How about having a separate doc that only talks about setting up minikube? And then in all the docs, we just add a link saying this is how this guide expects the environment to be setup?

So a doc covering all the aspects of:
* installing minikube (this is basic requirement)
* enabling ingress addon (if user wants `odo url` commands to work.)
* enabling OLM addon (if user wants `odo service` commands to work for Operators)
* enabling Service Catalog using Helm (if user wants `odo service` commands to work for Service Catalog.)

And then in doc about linking, we point the user to the sections of "minikube setup" doc that they need to refer as set of prerequisites. 

WDYT @boczkowska @kadel ?

**Comment by boczkowska** [comment]
@dharmit pushed the changes (minikube docs as a separate file and some edits in the language department), should be easier to review too.

File: docs/public/installing-and-onfiguring-minikube-environment.adoc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
++
+[source,sh]
+----
+$ kubectl create secret docker-registry regcred --docker-server=<your-registry-server> --docker-username=<your-name> --docker-password=<your-pword> --docker-email=<your-email>
+----
++
+where:
++
+* <your-registry-server> is the DockerHub Registry FQDN. (https://index.docker.io/v1/)
+* your-name> is your Docker username.
```
Comment (dharmit) [inline docs/public/installing-and-onfiguring-minikube-environment.adoc:70]: Missed an `<` here.

File: docs/public/installing-and-onfiguring-minikube-environment.adoc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+* <your-pword> is your Docker password.
+* <your-email> is your Docker email.
```
Comment (dharmit) [inline docs/public/installing-and-onfiguring-minikube-environment.adoc:72]: Nitpick - aren't these "Docker Hub" configurations? Something like "email used for Docker Hub account" would make more sense, maybe?

File: docs/public/installing-and-onfiguring-minikube-environment.adoc [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+. Add this new cred (`regcred` in the example above) to the default service account in minikube:
```
Comment (dharmit) [inline docs/public/installing-and-onfiguring-minikube-environment.adoc:74]: Please change `Add this new cred` to `Add this new secret`. Asking this because we're creating a secret in the previous step.

File: docs/public/installing-and-onfiguring-minikube-environment.adoc [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+[NOTE]
+----
+The guide only works with minikube configured with kubernetes 1.19.x or lower. `odo link` cannot link services successfully in a Kubernetes 1.20.x or newer environment.
+----
+
+It is recommended that users of this guide obtain a suitable system for running minikube with kubernetes. In practice this should be a 4 core system minimum. Before proceeding to the sample application, please follow the instructions for establishing a minikube environment:
+
+.Prerequisites
+* You have installed docker. See link:https://docs.docker.com/engine/install/[the Docker documentation].
```
Comment (dharmit) [inline docs/public/installing-and-onfiguring-minikube-environment.adoc:11]: My concern with Docker is that it is leading to a lot of other steps to be performed further in the documentation. I know that the documentation I linked earlier had this info and you have used it to create this one. But I am inclined towards asking the users to create a VM using minikube. This way, we can probably skip a few steps and ask the users to refer the minikube documentation instead.

TL;DR - I'm very apprehensive of spinning up minikube using docker driver.

WDYT @kadel?

File: docs/public/deploying-java-app-with-database.adoc [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+
+== Installation and Configuration of the minikube environment
+[NOTE]
+====
+The guide only works with minikube configured with kubernetes 1.19.x or lower. `odo link` cannot link services successfully in a Kubernetes 1.20.x or newer environment.
+====
+
+It is recommended that users of this guide obtain a suitable system for running minikube with kubernetes. In practice this should be a 4 core system minimum. Before proceeding to the sample application, please follow the instructions for establishing a minikube environment:
+
+=== Installing and starting docker
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:13]: This might need to be removed if we decide to go ahead with my idea of starting minikube as a VM instead of using docker driver.

File: docs/public/deploying-java-app-with-database.adoc [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+----
+
+==== Configuring minikube
+The application requires an ingress add-on to allow the routes to be created easily. Configure minikube for ingress by adding ingress as a minikube add-on:
+[source,sh]
+----
+> minikube addons enable ingress
+----
+
+It is possible that you face the Dockerhub pull rate limit if you do not have a pull secret for your personal free Dockerhub account. During ingress initialization two of the job pods used by ingress may fail to initialize due to pull rate limits. If this happens, and ingress fails to enable, you add a secret for the pulls for the following service accounts:
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:44]: Pull rate limit is appearing in both this and the minikube setup doc. We should have it at only place, no?

File: docs/public/deploying-java-app-with-database.adoc (hunk 1 above)
Comment (dharmit) [reply docs/public/deploying-java-app-with-database.adoc:110]: This is still an issue.

File: docs/public/deploying-java-app-with-database.adoc [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+** <your-registry-server> is the DockerHub Registry FQDN. (https://index.docker.io/v1/)
+** <your-name> is your Docker username.
+** <your-pword> is your Docker password.
+** <your-email> is your Docker email.
+
+* Add this new cred ('regcred' in the example above) to the default service account in minikube:
++
+[source,sh]
+----
+$ kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "regcred"}]}'
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:107]: Everything up to this point and my previous comment about "pull rate limiting appearing twice" is repeated in the two docs, IIUC.

File: docs/public/deploying-java-app-with-database.adoc [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+----
+
+Kubernetes Dashboard graphical UI config:<br>
+It is helpful to make use of the basic kubernetes dashboard UI to interact with the various kubernetes entities in a graphical way. Refer to the directions [here](https://minikube.sigs.k8s.io/docs/handbook/dashboard/) for enabling and starting the dashboard. Please note, this require the installation of and access to a desktop environment in order to make use of the dashboard. (GNOME + xrdb for example)
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:111]: TBH, we don't need Kubernetes Dashboard for odo related things. We can skip this.

File: docs/public/deploying-java-app-with-database.adoc [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+$ odo project set my-postgresql-operator-dev4devs-com
+----
+
+You can use the default configurations of the PostgreSQL Operator to start a Postgres database from it. But since our app uses few specific configuration values, lets make sure they are properly populated in the database service we start.
+
+. Store the YAML of the service in a file:
++
+[source,sh]
+----
+$ odo service create postgresql-operator.v0.1.1/Database --dry-run $ db.yaml
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:176]:
```suggestion
$ odo service create postgresql-operator.v0.1.1/Database --dry-run > db.yaml
```

File: docs/public/deploying-java-app-with-database.adoc [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+----
+
+. Modify and add following values under `metadata:`section in the `db.yaml` file:
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:179]:
```suggestion
. Modify and add following values under `metadata:` section in the `db.yaml` file:
```

File: docs/public/deploying-java-app-with-database.adoc [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+[source,sh]
+----
+$ odo push
+----
++
+After the link has been created and pushed a secret will have been created containing the database connection data that the application requires.
++
+You can inspect the new intermediate secret via the dashboard console in the 'my-postgresql-operator-dev4devs-com' namespace by navigating to Secrets and clicking on the secret named `mysboproj-database-sampledatabase` Notice it contains 4 pieces of data all related to the connection information for your PostgreSQL database instance.
++
+. Push the newly created link. This will terminate the existing application pod and start a new application pod.
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:374]: Not sure why we need this `odo push`. We already did `odo push` once after doing `odo link`. That should suffice. Another `odo push` is needed only if we're making any change to the application which, as far as I can see, we aren't.

@ajm01 any idea?

File: docs/public/deploying-java-app-with-database.adoc [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
++
+Once the new version is up (there will be a slight delay until application is available), navigate to the CreatePerson.xhtml using the URL created in a previous step. Enter requested data and click the *Save* button
+//![Create Person xhtml page](../../assets/createPersonDB.png)
++
+Notice you are re-directed to the `PersonList.xhtml` page, where your data is displayed having been input to the postgreSQL database and retrieved for display purposes.
+//![Create Person xhtml page](../../assets/displayPeople.png)
++
+You may inspect the database instance itself and query the table to see the data in place by using the postgreSQL command line tool, psql.
+
+. Navigate to the pod containing your db from the Openshift Console
```
Comment (dharmit) [inline docs/public/deploying-java-app-with-database.adoc:400]:
```suggestion
. Navigate to the pod containing your db from the Kubernetes Dashboard
```

**Review by dharmit — CHANGES_REQUESTED** [review]

File: docs/public/deploying-java-app-with-database.adoc (hunk 6 above)
Comment (boczkowska) [reply docs/public/deploying-java-app-with-database.adoc]: @dharmit right, I forgot to commit the changes in that Java doc. -_- Apologies, fixed.

**Comment by boczkowska** [comment]
@dharmit I'm a dimwit and I forgot to commit the changes to the java file yesterday (hence some duplication of the content). >:( Fixed now, also addressed your feedback.

**Review by dharmit — APPROVED** [review]
@boczkowska thanks for helping out ! :+1:

File: docs/public/deploying-java-app-with-database.adoc (hunk 12 above)
Comment (ajm01) [reply docs/public/deploying-java-app-with-database.adoc:248]: Hi @dharmit - apologies, I didn't see these updates in my private gmail ... i think that second odo push is extraneous...so not needed at that point in the doc.

Comment (dharmit) [reply docs/public/deploying-java-app-with-database.adoc:248]: @ajm01 OK, thanks. I'll remove that in an upcoming PR. :+1:
